### PR TITLE
Fix using enter on row, then hitting load causes exception in SANS

### LINKED ISSUE #31202: SANS Interface: Load throws an exception if you have input focus in a cell in a new row
Discovered testing Mantid Nightly 6.0.20210421.2323 on W10/x64. But probably exists in earlier versions. And probably affects all platforms.

image

If you hit Return/Enter after completing data entry in a row in the Reduction Table in the SANS Interface a new row is automatically created and input focus is given to the cell in this new row in the same column as the one in which Return/Enter was made.

If you click Load at this point - whilst the cell is still editable - an exception is immediately thrown:
```
Traceback (most recent call last):
  File "C:\MantidNightlyInstall\scripts\Interface\ui\sans_isis\sans_data_processor_gui.py", line 465, in _load_clicked
    self._call_settings_listeners(lambda listener: listener.on_load_clicked())
  File "C:\MantidNightlyInstall\scripts\Interface\ui\sans_isis\sans_data_processor_gui.py", line 244, in _call_settings_listeners
    target(listener)
  File "C:\MantidNightlyInstall\scripts\Interface\ui\sans_isis\sans_data_processor_gui.py", line 465, in <lambda>
    self._call_settings_listeners(lambda listener: listener.on_load_clicked())
  File "C:\MantidNightlyInstall\scripts\SANS\sans\gui_logic\presenter\run_tab_presenter.py", line 145, in on_load_clicked
    self._presenter.on_load_clicked()
  File "C:\MantidNightlyInstall\scripts\SANS\sans\gui_logic\presenter\run_tab_presenter.py", line 684, in on_load_clicked
    selected_rows = self._get_selected_non_empty_rows()
  File "C:\MantidNightlyInstall\scripts\SANS\sans\gui_logic\presenter\run_tab_presenter.py", line 655, in _get_selected_non_empty_rows
    to_process = [self._table_model.get_row(i) for i in selected_rows]
  File "C:\MantidNightlyInstall\scripts\SANS\sans\gui_logic\presenter\run_tab_presenter.py", line 655, in <listcomp>
    to_process = [self._table_model.get_row(i) for i in selected_rows]
  File "C:\MantidNightlyInstall\scripts\SANS\sans\gui_logic\models\table_model.py", line 57, in get_row
    return self._table_entries[index]
IndexError: list index out of range
```

This is surprisingly easy to trigger if, for example, you just want to load a new run. You type in a new run number, hit return and then immediately click Load, not realising you don't actually have the row you just typed in selected!

This should be gracefully trapped.

Because the SANS Reduction Table shares code with the Reflectometry Reduction Table, this issue is probably common to both.

### PR DESCRIPTION
**Pull request by DavidFair**
**Description of work.**
Fixes an issue where hitting enter then load will cause the SANS GUI to throw an unexpected exception. See each commit for more details

**Report to:** @smk78 

**To test:**
- Download the ISIS Training Dataset
- Ensure the LOQ folder is present, and in your data directory (so that LOQ74044 is present)
- In the SANS GUI set the user file to either the .txt or .toml file in that directory
- In the first column + first row enter 74044
- Hit enter, then immediately click load without doing anything else. It should correctly load row 0 instead of erroring

Fixes #31202

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by RichardWaiteSTFC** [comment]
As discussed offline (commenting here so we don't forget about this PR) after step 4 on windows I get the following error
```
Traceback (most recent call last):
 File "C:\mantid\scripts\Interface\ui\sans_isis\sans_data_processor_gui.py", line 488, in _append_and_edit_at_row_below_requested
 self._call_settings_listeners(lambda listener: listener.on_insert_row())
 File "C:\mantid\scripts\Interface\ui\sans_isis\sans_data_processor_gui.py", line 240, in _call_settings_listeners
 target(listener)
 File "C:\mantid\scripts\Interface\ui\sans_isis\sans_data_processor_gui.py", line 488, in <lambda>
 self._call_settings_listeners(lambda listener: listener.on_insert_row())
 File "C:\mantid\scripts\SANS\sans\gui_logic\presenter\run_tab_presenter.py", line 181, in on_insert_row
 self._presenter.on_insert_row()
 File "C:\mantid\scripts\SANS\sans\gui_logic\presenter\run_tab_presenter.py", line 815, in on_insert_row
 self.update_view_from_table_model()
 File "C:\mantid\scripts\SANS\sans\gui_logic\presenter\run_tab_presenter.py", line 503, in update_view_from_table_model
 self._view.clear_table()
 File "C:\mantid\scripts\Interface\ui\sans_isis\sans_data_processor_gui.py", line 2164, in clear_table
 self.data_processor_table.removeAllRows()
Exception: unknown
```

**Comment by DavidFair** [comment]
@RichardWaiteSTFC Can you still reproduce this? I've tried with Windows and Ubuntu and it works for both :thinking:

**Comment by DavidFair** [comment]
@RichardWaiteSTFC I managed to reproduce it locally, this should resolve it
In a nutshell, we need to append the row when the table widget asks us to, not after else memory gets out of sync

To be honest the entire MVP pattern is a mess in ISIS SANS at the moment; we have some places where the view triggers the presenter to query the view for details, other places where the view passes stuff through a listener and places where they go out of sync until we hit a critical task (like process) then sync up.
I need to sit down during the upcoming maintenance, create a design document for a unified approach and just apply it over the whole interface. But in the interim we'll have to stick with patchwork like this

File: scripts/Interface/ui/sans_isis/sans_data_processor_gui.py [hunk 1]
```diff
@@ -484,6 +480,7 @@ def _data_changed(self, row_location, column, old_value, new_value):
     def _row_inserted(self, row_location):
         if row_location.depth() > 1:
             self.data_processor_table.removeRowAt(row_location)
+        self._call_settings_listeners(lambda listener: listener.on_insert_row())
```
Comment (RichardWaiteSTFC) [inline scripts/Interface/ui/sans_isis/sans_data_processor_gui.py:483]: Looks good and fixes the bug - haven't seen this before, normally I would send a signal and that would be connected to a slot in the presenter - here it looks like you keep track of all the listeners and explicitly call a method of all listeners after an event - is there any advantage to this? One problem might be that you might want multiple listeners (but for different things) such that for example not all of them have a method called `on_insert_row()`

**Review by RichardWaiteSTFC — APPROVED** [review]

File: scripts/Interface/ui/sans_isis/sans_data_processor_gui.py (hunk 1 above)
Comment (RichardWaiteSTFC) [reply scripts/Interface/ui/sans_isis/sans_data_processor_gui.py:483]: As discussed offline @DavidFair is already planning to look at this further in maintenance. I'm very happy for this to get merged.

**Pull request merged**